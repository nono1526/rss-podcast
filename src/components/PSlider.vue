<template>
  <div
    class="p-slider"
    draggable
    @mousedown="mousedown"
  >
    <div
      class="p-slider__track-container"
    >
      <div
        class="p-slider__track-fill"
        :style="{
          width: `${precent}%`
        }"
      >
        <div
          v-if="!disabled"
          class="p-slider__thumb group-hover:block"
        />
      </div>
    </div>
  </div>
</template>

<script>
import { computed, onMounted, onUnmounted, ref } from 'vue'
export default {

  name: 'PSlider',
  inheritAttrs: false,
  props: {
    min: {
      type: Number,
      default: 0
    },
    max: {
      type: Number,
      default: 100
    },
    value: {
      type: Number,
      default: 50
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  emits: ['input', 'dragstart', 'dragging', 'dragend'],
  setup (props, { emit }) {
    const isMousedown = ref(false)
    let startX = 0
    let startOffsetX = 0
    let barWidth = 0
    const mousedown = e => {
      e.preventDefault()
      if (props.disabled) return false
      const { width } = e.currentTarget.getBoundingClientRect()
      barWidth = width
      isMousedown.value = true
      startX = e.clientX
      startOffsetX = e.clientX - e.currentTarget.offsetLeft
      internalValue.value = getValueByBarInteraction(startOffsetX, barWidth)
      emit('dragstart', e)
    }

    const moving = e => {
      if (!isMousedown.value) return false
      const dx = e.clientX - startX
      const offsetX = startOffsetX + dx
      internalValue.value = getValueByBarInteraction(offsetX, barWidth)
      emit('dragging', e)
    }

    const mouseup = e => {
      if (!isMousedown.value) return false
      isMousedown.value = false
      emit('dragend', e)
    }
    const precent = computed(() => {
      return ((props.value - props.min) / (props.max - props.min)) * 100
    })

    const internalValue = computed({
      get () {
        return valueRangeValidator(props.value)
      },
      set (val) {
        return emit('input', valueRangeValidator(val))
      }
    })

    const getValueByBarInteraction = (offsetX, barWidth) => {
      return Math.round((offsetX / barWidth) * props.max)
    }

    const valueRangeValidator = value => {
      value = Math.max(props.min, value)
      value = Math.min(props.max, value)
      return value
    }

    const bindEvents = () => {
      document.addEventListener('mousemove', moving)
      document.addEventListener('mouseup', mouseup)
    }

    const unBindEvents = () => {
      document.removeEventListener('mousemove', moving)
      document.removeEventListener('mouseup', mouseup)
    }

    onMounted(() => {
      bindEvents()
    })

    onUnmounted(() => {
      unBindEvents()
    })
    return {
      precent,
      mousedown,
      internalValue
    }
  }

}
</script>

<style>
.p-slider {
  @apply h-5 w-full flex items-center cursor-pointer group
}

.p-slider .p-slider__track-container {
  @apply rounded shadow bg-white relative h-1 flex-1
}

.p-slider .p-slider__thumb {
  @apply hidden h-3 w-3 rounded-full bg-blue-500 absolute shadow-2xl transform -translate-y-2/4 top-2/4 translate-x-2/4 right-0 cursor-pointer
}

.p-slider .p-slider__track-fill {
  @apply h-full bg-blue-400 rounded relative
}
</style>
