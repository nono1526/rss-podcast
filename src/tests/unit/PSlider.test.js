import PSlider from '@src/components/PSlider.vue'
import { shallowMount } from '@vue/test-utils'
describe('PSlider.vue', () => {
  let mountFunction
  beforeEach(() => {
    mountFunction = options => shallowMount(PSlider, {
      ...options
    })
  })

  it('should match snapshot', async () => {
    const wrapper = mountFunction({})

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('has correct track fill percent width with props.min/max/value', async () => {
    const wrapper = mountFunction({
      propsData: {
        min: 50,
        max: 100,
        value: 60
      }
    })
    const fill = wrapper.find('.p-slider__track-fill')
    expect(fill.element.style.width).toBe('20%')
    await wrapper.setProps({
      value: 50
    })
    expect(fill.element.style.width).toBe('0%')
    await wrapper.setProps({
      value: 100
    })
    expect(fill.element.style.width).toBe('100%')
    await wrapper.setProps({
      value: 100,
      min: 0,
      max: 100
    })
    expect(fill.element.style.width).toBe('100%')
    await wrapper.setProps({
      value: 50
    })
    expect(fill.element.style.width).toBe('50%')
  })

  it('mouse event flow', () => {
    const wrapper = mountFunction({})

    const input = wrapper.find('.p-slider')

    input.trigger('mousedown')

    expect(wrapper.emitted('dragstart')).toHaveLength(1)
    expect(wrapper.emitted('input')).toHaveLength(1)

    document.dispatchEvent(new Event('mousemove'))
    expect(wrapper.emitted('input')).toHaveLength(2)

    document.dispatchEvent(new Event('mouseup'))
    expect(wrapper.emitted('dragend')).toHaveLength(1)

    document.dispatchEvent(new Event('mousemove'))
    expect(wrapper.emitted('input')).toHaveLength(2) // not 3 because mouse is up
  })

  it('should not react to mousedown when disabled', () => {
    const wrapper = mountFunction({
      propsData: {
        disabled: true
      }
    })
    const input = wrapper.find('.p-slider')

    input.trigger('mousedown')
    expect(wrapper.emitted('dragstart')).toBeFalsy()
  })

  it('internalValue get correctly value by min/max', async () => {
    const wrapper = mountFunction({
      propsData: {
        min: 2,
        max: 5,
        value: 10
      }
    })
    expect(wrapper.vm.internalValue).toBe(5)
    await wrapper.setProps({
      value: 0
    })
    expect(wrapper.vm.internalValue).toBe(2)
  })

  it('internalValue set correctly value by min/max', async () => {
    const wrapper = mountFunction({
      propsData: {
        min: 2,
        max: 5,
        value: 10
      }
    })
    wrapper.vm.internalValue = 0
    expect(wrapper.emitted('input')).toEqual([[2]])
    wrapper.vm.internalValue = 10
    expect(wrapper.emitted('input')).toEqual([[2], [5]])
  })
})
