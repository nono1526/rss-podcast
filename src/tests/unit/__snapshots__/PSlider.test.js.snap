// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PSlider.vue should match snapshot 1`] = `
"<div class=\\"p-slider\\" draggable=\\"\\">
  <div class=\\"p-slider__track-container\\">
    <div class=\\"p-slider__track-fill\\" style=\\"width: 50%;\\">
      <div class=\\"p-slider__thumb group-hover:block\\"></div>
    </div>
  </div>
</div>"
`;
